// @microsoft/agents-activity/dist/src/activityTypes.js
"use strict";
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.activityTypesZodSchema = exports.ActivityTypes = void 0;
const zod_1 = require("zod");
/**
 * Enum representing activity types.
 */
var ActivityTypes;
(function (ActivityTypes) {
    ActivityTypes["Message"] = "message";
    ActivityTypes["ContactRelationUpdate"] = "contactRelationUpdate";
    ActivityTypes["ConversationUpdate"] = "conversationUpdate";
    ActivityTypes["Typing"] = "typing";
    ActivityTypes["EndOfConversation"] = "endOfConversation";
    ActivityTypes["Event"] = "event";
    ActivityTypes["Invoke"] = "invoke";
    ActivityTypes["InvokeResponse"] = "invokeResponse";
    ActivityTypes["DeleteUserData"] = "deleteUserData";
    ActivityTypes["MessageUpdate"] = "messageUpdate";
    ActivityTypes["MessageDelete"] = "messageDelete";
    ActivityTypes["InstallationUpdate"] = "installationUpdate";
    ActivityTypes["MessageReaction"] = "messageReaction";
    ActivityTypes["Suggestion"] = "suggestion";
    ActivityTypes["Trace"] = "trace";
    ActivityTypes["Handoff"] = "handoff";
    ActivityTypes["Command"] = "command";
    ActivityTypes["CommandResult"] = "commandResult";
    ActivityTypes["Delay"] = "delay";
})(ActivityTypes || (exports.ActivityTypes = ActivityTypes = {}));
/**
 * Zod schema for validating an ActivityTypes enum.
 */
exports.activityTypesZodSchema = zod_1.z.enum([
    'message',
    'contactRelationUpdate',
    'conversationUpdate',
    'typing',
    'endOfConversation',
    'event',
    'invoke',
    'invokeResponse',
    'deleteUserData',
    'messageUpdate',
    'messageDelete',
    'installationUpdate',
    'messageReaction',
    'suggestion',
    'trace',
    'handoff',
    'command',
    'commandResult',
    'delay'
]);
