// @microsoft/agents-activity/dist/src/index.js
"use strict";
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextFormatTypes = exports.MessageReactionTypes = exports.InputHints = exports.DeliveryModes = exports.CallerIdConstants = exports.ActivityTypes = exports.ActivityImportance = exports.ActivityEventNames = exports.activityZodSchema = exports.Activity = exports.RoleTypes = exports.EndOfConversationCodes = exports.Channels = exports.AttachmentLayoutTypes = exports.SemanticActionStateTypes = exports.ActionTypes = void 0;
/**
 * Export statements for various modules.
 */
var actionTypes_1 = require("./action/actionTypes");
Object.defineProperty(exports, "ActionTypes", { enumerable: true, get: function () { return actionTypes_1.ActionTypes; } });
var semanticActionStateTypes_1 = require("./action/semanticActionStateTypes");
Object.defineProperty(exports, "SemanticActionStateTypes", { enumerable: true, get: function () { return semanticActionStateTypes_1.SemanticActionStateTypes; } });
var attachmentLayoutTypes_1 = require("./attachment/attachmentLayoutTypes");
Object.defineProperty(exports, "AttachmentLayoutTypes", { enumerable: true, get: function () { return attachmentLayoutTypes_1.AttachmentLayoutTypes; } });
var channels_1 = require("./conversation/channels");
Object.defineProperty(exports, "Channels", { enumerable: true, get: function () { return channels_1.Channels; } });
var endOfConversationCodes_1 = require("./conversation/endOfConversationCodes");
Object.defineProperty(exports, "EndOfConversationCodes", { enumerable: true, get: function () { return endOfConversationCodes_1.EndOfConversationCodes; } });
var roleTypes_1 = require("./conversation/roleTypes");
Object.defineProperty(exports, "RoleTypes", { enumerable: true, get: function () { return roleTypes_1.RoleTypes; } });
__exportStar(require("./invoke/adaptiveCardInvokeAction"), exports);
var activity_1 = require("./activity");
Object.defineProperty(exports, "Activity", { enumerable: true, get: function () { return activity_1.Activity; } });
Object.defineProperty(exports, "activityZodSchema", { enumerable: true, get: function () { return activity_1.activityZodSchema; } });
var activityEventNames_1 = require("./activityEventNames");
Object.defineProperty(exports, "ActivityEventNames", { enumerable: true, get: function () { return activityEventNames_1.ActivityEventNames; } });
var activityImportance_1 = require("./activityImportance");
Object.defineProperty(exports, "ActivityImportance", { enumerable: true, get: function () { return activityImportance_1.ActivityImportance; } });
var activityTypes_1 = require("./activityTypes");
Object.defineProperty(exports, "ActivityTypes", { enumerable: true, get: function () { return activityTypes_1.ActivityTypes; } });
var callerIdConstants_1 = require("./callerIdConstants");
Object.defineProperty(exports, "CallerIdConstants", { enumerable: true, get: function () { return callerIdConstants_1.CallerIdConstants; } });
var deliveryModes_1 = require("./deliveryModes");
Object.defineProperty(exports, "DeliveryModes", { enumerable: true, get: function () { return deliveryModes_1.DeliveryModes; } });
var inputHints_1 = require("./inputHints");
Object.defineProperty(exports, "InputHints", { enumerable: true, get: function () { return inputHints_1.InputHints; } });
var messageReactionTypes_1 = require("./messageReactionTypes");
Object.defineProperty(exports, "MessageReactionTypes", { enumerable: true, get: function () { return messageReactionTypes_1.MessageReactionTypes; } });
var textFormatTypes_1 = require("./textFormatTypes");
Object.defineProperty(exports, "TextFormatTypes", { enumerable: true, get: function () { return textFormatTypes_1.TextFormatTypes; } });
