// @fluentui/react-motion-components-preview/lib-commonjs/components/Rotate/Rotate.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Rotate", {
    enumerable: true,
    get: function() {
        return Rotate;
    }
});
const _reactmotion = require("@fluentui/react-motion");
const _fadeatom = require("../../atoms/fade-atom");
const _rotateatom = require("../../atoms/rotate-atom");
/**
 * Define a presence motion for rotate in/out
 *
 * @param duration - Time (ms) for the enter transition (rotate-in). Defaults to the `durationGentle` value.
 * @param easing - Easing curve for the enter transition (rotate-in). Defaults to the `curveDecelerateMax` value.
 * @param exitDuration - Time (ms) for the exit transition (rotate-out). Defaults to the `duration` param for symmetry.
 * @param exitEasing - Easing curve for the exit transition (rotate-out). Defaults to the `curveAccelerateMax` value.
 * @param axis - The axis of rotation: 'x', 'y', or 'z'. Defaults to 'y'.
 * @param angle - The starting rotation angle in degrees. Defaults to -90.
 * @param exitAngle - The ending rotation angle in degrees. Defaults to the negation of `angle`.
 * @param animateOpacity - Whether to animate the opacity during the rotation. Defaults to `true`.
 */ const rotatePresenceFn = ({ axis = 'y', angle = -90, exitAngle = -angle, duration = _reactmotion.motionTokens.durationGentle, exitDuration = duration, easing = _reactmotion.motionTokens.curveDecelerateMax, exitEasing = _reactmotion.motionTokens.curveAccelerateMax, animateOpacity = true })=>{
    const enterAtoms = [
        (0, _rotateatom.rotateAtom)({
            direction: 'enter',
            duration,
            easing,
            axis,
            angle,
            exitAngle
        })
    ];
    const exitAtoms = [
        (0, _rotateatom.rotateAtom)({
            direction: 'exit',
            duration: exitDuration,
            easing: exitEasing,
            axis,
            angle,
            exitAngle
        })
    ];
    if (animateOpacity) {
        enterAtoms.push((0, _fadeatom.fadeAtom)({
            direction: 'enter',
            duration,
            easing
        }));
        exitAtoms.push((0, _fadeatom.fadeAtom)({
            direction: 'exit',
            duration: exitDuration,
            easing: exitEasing
        }));
    }
    return {
        enter: enterAtoms,
        exit: exitAtoms
    };
};
const Rotate = (0, _reactmotion.createPresenceComponent)(rotatePresenceFn);
