// @microsoft/agents-hosting-teams/dist/src/app/messaging-extension/messageExtensionsInvokeNames.js
"use strict";
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageExtensionsInvokeNames = void 0;
var MessageExtensionsInvokeNames;
(function (MessageExtensionsInvokeNames) {
    MessageExtensionsInvokeNames["ANONYMOUS_QUERY_LINK_INVOKE"] = "composeExtension/anonymousQueryLink";
    MessageExtensionsInvokeNames["FETCH_TASK_INVOKE"] = "composeExtension/fetchTask";
    MessageExtensionsInvokeNames["QUERY_INVOKE"] = "composeExtension/query";
    MessageExtensionsInvokeNames["QUERY_LINK_INVOKE"] = "composeExtension/queryLink";
    MessageExtensionsInvokeNames["SELECT_ITEM_INVOKE"] = "composeExtension/selectItem";
    MessageExtensionsInvokeNames["SUBMIT_ACTION_INVOKE"] = "composeExtension/submitAction";
    MessageExtensionsInvokeNames["QUERY_SETTING_URL"] = "composeExtension/querySettingUrl";
    MessageExtensionsInvokeNames["CONFIGURE_SETTINGS"] = "composeExtension/setting";
    MessageExtensionsInvokeNames["QUERY_CARD_BUTTON_CLICKED"] = "composeExtension/onCardButtonClicked";
})(MessageExtensionsInvokeNames || (exports.MessageExtensionsInvokeNames = MessageExtensionsInvokeNames = {}));
