// @microsoft/agents-hosting-teams/dist/src/channel-data/teamsChannelData.js
"use strict";
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.teamsChannelDataZodSchema = void 0;
const zod_1 = require("zod");
const channelInfo_1 = require("./channelInfo");
const notificationInfo_1 = require("./notificationInfo");
const onBehalfOf_1 = require("./onBehalfOf");
const teamsChannelDataSettings_1 = require("./teamsChannelDataSettings");
const teamsMeetingInfo_1 = require("./teamsMeetingInfo");
const tenantInfo_1 = require("./tenantInfo");
const teamInfo_1 = require("./teamInfo");
/**
 * Zod schema for validating TeamsChannelData objects.
 */
exports.teamsChannelDataZodSchema = zod_1.z.object({
    channel: channelInfo_1.channelInfoZodSchema.optional(),
    eventType: zod_1.z.string().min(1).optional(),
    team: teamInfo_1.teamInfoZodSchema.optional(),
    notification: notificationInfo_1.notificationInfoZodSchema.optional(),
    tenant: tenantInfo_1.tenantInfoZodSchema.optional(),
    meeting: teamsMeetingInfo_1.teamsMeetingInfoZodSchema.optional(),
    settings: teamsChannelDataSettings_1.teamsChannelDataSettingsZodSchema.optional(),
    onBehalfOf: zod_1.z.array(onBehalfOf_1.onBehalfOfZodSchema).optional()
});
