// @microsoft/teamsfx-react/build/cjs/index.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseWidget = exports.BaseDashboard = exports.useGraphWithCredential = exports.useTeamsUserCredential = exports.useData = exports.useTeams = void 0;
var useTeams_1 = require("./useTeams");
Object.defineProperty(exports, "useTeams", { enumerable: true, get: function () { return useTeams_1.useTeams; } });
var useData_1 = require("./useData");
Object.defineProperty(exports, "useData", { enumerable: true, get: function () { return useData_1.useData; } });
var useTeamsUserCredential_1 = require("./useTeamsUserCredential");
Object.defineProperty(exports, "useTeamsUserCredential", { enumerable: true, get: function () { return useTeamsUserCredential_1.useTeamsUserCredential; } });
var useGraph_1 = require("./useGraph");
Object.defineProperty(exports, "useGraphWithCredential", { enumerable: true, get: function () { return useGraph_1.useGraphWithCredential; } });
var BaseDashboard_1 = require("./BaseDashboard");
Object.defineProperty(exports, "BaseDashboard", { enumerable: true, get: function () { return BaseDashboard_1.BaseDashboard; } });
var BaseWidget_1 = require("./BaseWidget");
Object.defineProperty(exports, "BaseWidget", { enumerable: true, get: function () { return BaseWidget_1.BaseWidget; } });
