// @microsoft/agents-hosting-dialogs/dist/src/dialogsComponentRegistration.js
"use strict";
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DialogsComponentRegistration = void 0;
const componentRegistration_1 = require("./componentRegistration");
const serviceCollection_1 = require("./serviceCollection");
const configuration_1 = require("./configuration");
const dialogsAgentComponent_1 = require("./dialogsAgentComponent");
/**
 * Makes dialogs component available to the system registering functionality.
 */
class DialogsComponentRegistration extends componentRegistration_1.ComponentRegistration {
    /**
       * Creates an instance of the DialogsComponentRegistration class.
       */
    constructor() {
        super();
        this.services = new serviceCollection_1.ServiceCollection({
            memoryScopes: [],
            pathResolvers: [],
        });
        new dialogsAgentComponent_1.DialogsAgentComponent().configureServices(this.services, configuration_1.noOpConfiguration);
    }
    /**
       * Gets the dialogs memory scopes.
       *
       * @returns {MemoryScope[]} A list of MemoryScope.
       */
    getMemoryScopes() {
        return this.services.mustMakeInstance('memoryScopes');
    }
    /**
       * Gets the dialogs path resolvers.
       *
       * @returns {PathResolver[]} A list of PathResolver.
       */
    getPathResolvers() {
        return this.services.mustMakeInstance('pathResolvers');
    }
}
exports.DialogsComponentRegistration = DialogsComponentRegistration;
