// @microsoft/agents-hosting-dialogs/dist/src/dialogTurnStatus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DialogTurnStatus = void 0;
/**
 * Represents the state of the dialog stack after a dialog context attempts to begin, continue,
 * or otherwise manipulate one or more dialogs.
 *
 */
var DialogTurnStatus;
(function (DialogTurnStatus) {
    /**
       * The dialog stack is empty.
       *
       * @remarks
       * Indicates that the dialog stack was initially empty when the operation was attempted.
       */
    DialogTurnStatus["empty"] = "empty";
    /**
       * The active dialog on top of the stack is waiting for a response from the user.
       */
    DialogTurnStatus["waiting"] = "waiting";
    /**
       * The last dialog on the stack completed successfully.
       *
       * @remarks
       * Indicates that a result might be available and the stack is now empty.
       *
       */
    DialogTurnStatus["complete"] = "complete";
    /**
       * All dialogs on the stack were cancelled and the stack is empty.
       */
    DialogTurnStatus["cancelled"] = "cancelled";
    /**
       * Current dialog completed successfully, but turn should end.
       */
    DialogTurnStatus["completeAndWait"] = "completeAndWait";
})(DialogTurnStatus || (exports.DialogTurnStatus = DialogTurnStatus = {}));
