// @microsoft/agents-activity/dist/src/action/actionTypes.js
"use strict";
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.actionTypesZodSchema = exports.ActionTypes = void 0;
const zod_1 = require("zod");
/**
 * Enum representing the types of actions.
 */
var ActionTypes;
(function (ActionTypes) {
    ActionTypes["OpenUrl"] = "openUrl";
    ActionTypes["ImBack"] = "imBack";
    ActionTypes["PostBack"] = "postBack";
    ActionTypes["PlayAudio"] = "playAudio";
    ActionTypes["PlayVideo"] = "playVideo";
    ActionTypes["ShowImage"] = "showImage";
    ActionTypes["DownloadFile"] = "downloadFile";
    ActionTypes["Signin"] = "signin";
    ActionTypes["Call"] = "call";
    ActionTypes["MessageBack"] = "messageBack";
    ActionTypes["OpenApp"] = "openApp";
})(ActionTypes || (exports.ActionTypes = ActionTypes = {}));
/**
 * Zod schema for validating ActionTypes.
 */
exports.actionTypesZodSchema = zod_1.z.enum(['openUrl', 'imBack', 'postBack', 'playAudio', 'showImage', 'downloadFile', 'signin', 'call', 'messageBack', 'openApp']);
