// @microsoft/agents-hosting-dialogs/dist/src/memory/scopes/memoryScope.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MemoryScope = void 0;
/**
 * Abstract base class for all memory scopes.
 */
class MemoryScope {
    /**
       * Initializes a new instance of the MemoryScope class.
       *
       * @param name Name of the scope.
       * @param includeInSnapshot Boolean value indicating whether this memory
       * should be included in snapshot. Default value is true.
       */
    constructor(name, includeInSnapshot = true) {
        this.includeInSnapshot = includeInSnapshot;
        this.name = name;
    }
    /**
       * Changes the backing object for the memory scope.
       *
       * @param _dialogContext Current dialog context
       * @param _memory memory to assign
       */
    setMemory(_dialogContext, _memory) {
        throw new Error(`MemoryScope.setMemory: The '${this.name}' memory scope is read-only.`);
    }
    /**
       * Loads a scopes backing memory at the start of a turn.
       *
       * @param _dialogContext Current dialog context.
       */
    async load(_dialogContext) {
        // No initialization by default.
    }
    /**
       * Saves a scopes backing memory at the end of a turn.
       *
       * @param _dialogContext Current dialog context.
       */
    async saveChanges(_dialogContext) {
        // No initialization by default.
    }
    /**
       * Deletes the backing memory for a scope.
       *
       * @param _dialogContext Current dialog context.
       */
    async delete(_dialogContext) {
        throw new Error(`MemoryScope.delete: The '${this.name}' memory scope can't be deleted.`);
    }
}
exports.MemoryScope = MemoryScope;
